src/Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAl.cpp

```diff
@@ -0,0 +1,342 @@
+        compute_r_bar_squared(mu, x, r_squared_, r_dot_b_squared_);
+    // Equation (49)
+    return mu * sqrt(square(h_0_) + r_bar_squared) - 1.0;
+  }
+
+ private:
+  const double h_0_;
+  const double r_squared_;
+  const double b_squared_;
+  const double r_dot_b_squared_;
+};
+
+// Master function, see Sec. II.E
+template <size_t ThermodynamicDim>
+class FunctionOfMu {
+ public:
+  FunctionOfMu(const double total_energy_density,
+               const double momentum_density_squared,
+               const double momentum_density_dot_magnetic_field,
+               const double magnetic_field_squared,
+               const double rest_mass_density_times_lorentz_factor,
+               const EquationsOfState::EquationOfState<true, ThermodynamicDim>&
+                   equation_of_state) noexcept
+      : q_(total_energy_density / rest_mass_density_times_lorentz_factor - 1.0),
+        r_squared_(momentum_density_squared /
+                   square(rest_mass_density_times_lorentz_factor)),
+        b_squared_(magnetic_field_squared /
+                   rest_mass_density_times_lorentz_factor),
+        r_dot_b_squared_(square(momentum_density_dot_magnetic_field) /
+                         cube(rest_mass_density_times_lorentz_factor)),
+        rest_mass_density_times_lorentz_factor_(
+            rest_mass_density_times_lorentz_factor),
+        equation_of_state_(equation_of_state),
+        h_0_(equation_of_state_.specific_enthalpy_lower_bound()),
+        v_0_squared_(compute_v_0_squared(r_squared_, h_0_)) {}
+
+  std::pair<double, double> root_bracket(
+      const double rest_mass_density_times_lorentz_factor,
+      const double absolute_tolerance, const double relative_tolerance,
+      const size_t max_iterations) const {
+    // see text between Equations (49) and (50) and after Equation (54)
+    double lower_bound = 0.0;
+    double upper_bound = 1.0 / h_0_;
+    if (r_squared_ < square(h_0_)) {
+      // need to solve auxiliary function to determine mu_+ which will
+      // be the upper bound for the master function bracket
+      const auto auxiliary_function =
+          AuxiliaryFunction{h_0_, r_squared_, b_squared_, r_dot_b_squared_};
+      try {
+        upper_bound =
+            // NOLINTNEXTLINE(clang-analyzer-core)
+            RootFinder::toms748(auxiliary_function, lower_bound, upper_bound,
+                                absolute_tolerance, relative_tolerance,
+                                max_iterations);
+      } catch (std::exception& exception) {
+        throw;
+      }
+    }
+
+    // Determine if the corner case discussed in Appendix A occurs where the
+    // mass density is outside the valid range of the EOS
+    const double rho_min = equation_of_state_.rest_mass_density_lower_bound();
+    const double rho_max = equation_of_state_.rest_mass_density_upper_bound();
+
+    if (rest_mass_density_times_lorentz_factor < rho_min) {
+      throw std::runtime_error("Density too small for EOS");
```

Comment:
could you add a code comment and/or a doxygen comment on what to do if one hits this case?